File: backend/app/core/agenda_config.py
"""Helpers para configuracao de funcionamento da agenda."""
from __future__ import annotations

import json
from datetime import date, datetime
from typing import Any

DIA_SEMANA_KEYS = ["1", "2", "3", "4", "5", "6", "7"]

NOMES_DIA_SEMANA = {
    "1": "segunda-feira",
    "2": "terca-feira",
    "3": "quarta-feira",
    "4": "quinta-feira",
    "5": "sexta-feira",
    "6": "sabado",
    "7": "domingo",
}

DEFAULT_AGENDA_SEMANAL = {
    "1": {"ativo": True, "inicio": "08:00", "fim": "14:00"},
    "2": {"ativo": True, "inicio": "08:00", "fim": "14:00"},
    "3": {"ativo": True, "inicio": "08:00", "fim": "14:00"},
    "4": {"ativo": True, "inicio": "08:00", "fim": "14:00"},
    "5": {"ativo": True, "inicio": "08:00", "fim": "14:00"},
    "6": {"ativo": True, "inicio": "09:00", "fim": "13:00"},
    "7": {"ativo": False, "inicio": "09:00", "fim": "13:00"},
}

DEFAULT_EXCECAO_INICIO = "08:00"
DEFAULT_EXCECAO_FIM = "18:00"


def _normalizar_hora_hhmm(value: Any, fallback: str) -> str:
    if not isinstance(value, str):
        return fallback
    raw = value.strip()
    if len(raw) != 5 or raw[2] != ":":
        return fallback

    hora_str = raw[:2]
    minuto_str = raw[3:]
    if not (hora_str.isdigit() and minuto_str.isdigit()):
        return fallback

    hora = int(hora_str)
    minuto = int(minuto_str)
    if hora < 0 or hora > 23 or minuto < 0 or minuto > 59:
        return fallback
    return f"{hora:02d}:{minuto:02d}"


def _hora_em_minutos(value: str) -> int:
    hora_str, minuto_str = value.split(":")
    return int(hora_str) * 60 + int(minuto_str)


def carregar_json(raw: Any) -> Any:
    if raw is None:
        return None
    if isinstance(raw, (dict, list)):
        return raw
    if not isinstance(raw, str):
        return None

    texto = raw.strip()
    if not texto:
        return None

    try:
        return json.loads(texto)
    except json.JSONDecodeError:
        return None


def normalizar_agenda_semanal(payload: Any) -> dict[str, dict[str, Any]]:
    source = payload if isinstance(payload, dict) else {}
    agenda: dict[str, dict[str, Any]] = {}

    for dia in DIA_SEMANA_KEYS:
        default = DEFAULT_AGENDA_SEMANAL[dia]
        item = source.get(dia) if isinstance(source.get(dia), dict) else {}

        ativo = item.get("ativo", default["ativo"])
        inicio = _normalizar_hora_hhmm(item.get("inicio"), default["inicio"])
        fim = _normalizar_hora_hhmm(item.get("fim"), default["fim"])

        if _hora_em_minutos(inicio) >= _hora_em_minutos(fim):
            inicio = default["inicio"]
            fim = default["fim"]

        agenda[dia] = {
            "ativo": bool(ativo),
            "inicio": inicio,
            "fim": fim,
        }

    return agenda


def normalizar_agenda_feriados(payload: Any) -> list[dict[str, str]]:
    if not isinstance(payload, list):
        return []

    itens: list[dict[str, str]] = []
    datas_vistas: set[str] = set()

    for item in payload:
        if isinstance(item, str):
            data_raw = item.strip()
            descricao_raw = ""
            tipo_raw = "local"
        elif isinstance(item, dict):
            data_raw = str(item.get("data", "")).strip()
            descricao_raw = str(item.get("descricao", "")).strip()
            tipo_raw = str(item.get("tipo", "local")).strip().lower()
        else:
            continue

        if not data_raw:
            continue

        try:
            data_iso = datetime.strptime(data_raw, "%Y-%m-%d").date().isoformat()
        except ValueError:
            continue

        if data_iso in datas_vistas:
            continue
        datas_vistas.add(data_iso)

        tipo = "nacional" if tipo_raw == "nacional" else "local"
        itens.append(
            {
                "data": data_iso,
                "descricao": descricao_raw,
                "tipo": tipo,
            }
        )

    itens.sort(key=lambda row: row["data"])
    return itens


def normalizar_agenda_excecoes(payload: Any) -> list[dict[str, Any]]:
    if not isinstance(payload, list):
        return []

    mapa_itens: dict[str, dict[str, Any]] = {}

    for item in payload:
        if not isinstance(item, dict):
            continue

        data_raw = str(item.get("data", "")).strip()
        if not data_raw:
            continue

        try:
            data_iso = datetime.strptime(data_raw, "%Y-%m-%d").date().isoformat()
        except ValueError:
            continue

        ativo = bool(item.get("ativo", True))
        inicio = _normalizar_hora_hhmm(item.get("inicio"), DEFAULT_EXCECAO_INICIO)
        fim = _normalizar_hora_hhmm(item.get("fim"), DEFAULT_EXCECAO_FIM)
        if _hora_em_minutos(inicio) >= _hora_em_minutos(fim):
            inicio = DEFAULT_EXCECAO_INICIO
            fim = DEFAULT_EXCECAO_FIM

        motivo = str(item.get("motivo", "") or "").strip()

        mapa_itens[data_iso] = {
            "data": data_iso,
            "ativo": ativo,
            "inicio": inicio,
            "fim": fim,
            "motivo": motivo,
        }

    itens = list(mapa_itens.values())
    itens.sort(key=lambda row: row["data"])
    return itens


def carregar_agenda_semanal(raw: Any) -> dict[str, dict[str, Any]]:
    payload = carregar_json(raw)
    return normalizar_agenda_semanal(payload)


def carregar_agenda_feriados(raw: Any) -> list[dict[str, str]]:
    payload = carregar_json(raw)
    return normalizar_agenda_feriados(payload)


def carregar_agenda_excecoes(raw: Any) -> list[dict[str, Any]]:
    payload = carregar_json(raw)
    return normalizar_agenda_excecoes(payload)


def serializar_json(payload: Any) -> str:
    return json.dumps(payload, ensure_ascii=False)


def obter_feriado(data_ref: date, agenda_feriados: list[dict[str, str]]) -> dict[str, str] | None:
    iso = data_ref.isoformat()
    for item in agenda_feriados:
        if item.get("data") == iso:
            return item
    return None


def obter_excecao_data(data_ref: date, agenda_excecoes: list[dict[str, Any]]) -> dict[str, Any] | None:
    iso = data_ref.isoformat()
    for item in agenda_excecoes:
        if item.get("data") == iso:
            return item
    return None


def validar_horario_agenda(
    inicio_local: datetime,
    fim_local: datetime,
    agenda_semanal: dict[str, dict[str, Any]],
    agenda_feriados: list[dict[str, str]],
    agenda_excecoes: list[dict[str, Any]] | None = None,
) -> tuple[bool, str]:
    if fim_local <= inicio_local:
        return False, "Horario invalido: o fim precisa ser maior que o inicio."

    if inicio_local.date() != fim_local.date():
        return False, "Agendamento deve comecar e terminar no mesmo dia."

    excecoes = agenda_excecoes or []
    excecao = obter_excecao_data(inicio_local.date(), excecoes)
    if excecao is not None:
        if not bool(excecao.get("ativo", False)):
            motivo = str(excecao.get("motivo") or "").strip()
            complemento = f" ({motivo})" if motivo else ""
            return False, f"Agenda fechada por excecao de data{complemento}."
        inicio_janela = _normalizar_hora_hhmm(excecao.get("inicio"), DEFAULT_EXCECAO_INICIO)
        fim_janela = _normalizar_hora_hhmm(excecao.get("fim"), DEFAULT_EXCECAO_FIM)
    else:
        feriado = obter_feriado(inicio_local.date(), agenda_feriados)
        if feriado:
            descricao = str(feriado.get("descricao") or "").strip()
            complemento = f" ({descricao})" if descricao else ""
            return False, f"Agenda fechada em feriado{complemento}."

        dia_key = str(inicio_local.isoweekday())
        dia_cfg = agenda_semanal.get(dia_key) or DEFAULT_AGENDA_SEMANAL[dia_key]

        if not bool(dia_cfg.get("ativo", False)):
            nome_dia = NOMES_DIA_SEMANA.get(dia_key, "este dia")
            return False, f"Agenda fechada para {nome_dia}."

        inicio_janela = _normalizar_hora_hhmm(dia_cfg.get("inicio"), DEFAULT_AGENDA_SEMANAL[dia_key]["inicio"])
        fim_janela = _normalizar_hora_hhmm(dia_cfg.get("fim"), DEFAULT_AGENDA_SEMANAL[dia_key]["fim"])

    inicio_janela_min = _hora_em_minutos(inicio_janela)
    fim_janela_min = _hora_em_minutos(fim_janela)
    inicio_ag_min = inicio_local.hour * 60 + inicio_local.minute
    fim_ag_min = fim_local.hour * 60 + fim_local.minute

    if inicio_ag_min < inicio_janela_min or fim_ag_min > fim_janela_min:
        if excecao is not None:
            return (
                False,
                f"Horario fora da excecao desta data ({inicio_janela} as {fim_janela}).",
            )
        return (
            False,
            f"Horario fora do funcionamento da agenda ({inicio_janela} as {fim_janela}).",
        )

    return True, ""

File: frontend/lib/agenda-config.ts
export type DiaSemanaKey = "1" | "2" | "3" | "4" | "5" | "6" | "7";

export interface AgendaDiaConfig {
  ativo: boolean;
  inicio: string;
  fim: string;
}

export interface AgendaFeriadoConfig {
  data: string;
  descricao?: string;
  tipo?: "local" | "nacional";
}

export interface AgendaExcecaoConfig {
  data: string;
  ativo: boolean;
  inicio: string;
  fim: string;
  motivo?: string;
}

export type AgendaSemanalConfig = Record<DiaSemanaKey, AgendaDiaConfig>;

export const DIA_SEMANA_KEYS: DiaSemanaKey[] = ["1", "2", "3", "4", "5", "6", "7"];

export const DIAS_SEMANA_LABELS: Array<{ id: DiaSemanaKey; nome: string }> = [
  { id: "1", nome: "Segunda-feira" },
  { id: "2", nome: "Terca-feira" },
  { id: "3", nome: "Quarta-feira" },
  { id: "4", nome: "Quinta-feira" },
  { id: "5", nome: "Sexta-feira" },
  { id: "6", nome: "Sabado" },
  { id: "7", nome: "Domingo" },
];

export const DEFAULT_AGENDA_SEMANAL: AgendaSemanalConfig = {
  "1": { ativo: true, inicio: "08:00", fim: "14:00" },
  "2": { ativo: true, inicio: "08:00", fim: "14:00" },
  "3": { ativo: true, inicio: "08:00", fim: "14:00" },
  "4": { ativo: true, inicio: "08:00", fim: "14:00" },
  "5": { ativo: true, inicio: "08:00", fim: "14:00" },
  "6": { ativo: true, inicio: "09:00", fim: "13:00" },
  "7": { ativo: false, inicio: "09:00", fim: "13:00" },
};

const toDateInput = (date: Date) => {
  const ano = date.getFullYear();
  const mes = String(date.getMonth() + 1).padStart(2, "0");
  const dia = String(date.getDate()).padStart(2, "0");
  return `${ano}-${mes}-${dia}`;
};

export const normalizarHoraHHMM = (valor: unknown, fallback: string): string => {
  if (typeof valor !== "string") return fallback;
  const texto = valor.trim();
  const match = texto.match(/^(\d{2}):(\d{2})$/);
  if (!match) return fallback;
  const hora = Number.parseInt(match[1], 10);
  const minuto = Number.parseInt(match[2], 10);
  if (!Number.isFinite(hora) || !Number.isFinite(minuto) || hora < 0 || hora > 23 || minuto < 0 || minuto > 59) {
    return fallback;
  }
  return `${String(hora).padStart(2, "0")}:${String(minuto).padStart(2, "0")}`;
};

export const horarioParaMinutos = (horario: string): number | null => {
  const match = horario.match(/^(\d{2}):(\d{2})$/);
  if (!match) return null;
  const hora = Number.parseInt(match[1], 10);
  const minuto = Number.parseInt(match[2], 10);
  if (!Number.isFinite(hora) || !Number.isFinite(minuto)) return null;
  if (hora < 0 || hora > 23 || minuto < 0 || minuto > 59) return null;
  return hora * 60 + minuto;
};

export const normalizarAgendaSemanal = (payload: unknown): AgendaSemanalConfig => {
  const source = typeof payload === "object" && payload !== null ? (payload as Record<string, unknown>) : {};
  const agenda = {} as AgendaSemanalConfig;

  for (const dia of DIA_SEMANA_KEYS) {
    const defaultDia = DEFAULT_AGENDA_SEMANAL[dia];
    const rawDia = source[dia];
    const item = typeof rawDia === "object" && rawDia !== null ? (rawDia as Record<string, unknown>) : {};

    const inicio = normalizarHoraHHMM(item.inicio, defaultDia.inicio);
    const fim = normalizarHoraHHMM(item.fim, defaultDia.fim);

    const inicioMin = horarioParaMinutos(inicio) ?? horarioParaMinutos(defaultDia.inicio) ?? 0;
    const fimMin = horarioParaMinutos(fim) ?? horarioParaMinutos(defaultDia.fim) ?? 0;

    agenda[dia] = {
      ativo: typeof item.ativo === "boolean" ? item.ativo : defaultDia.ativo,
      inicio: inicioMin < fimMin ? inicio : defaultDia.inicio,
      fim: inicioMin < fimMin ? fim : defaultDia.fim,
    };
  }

  return agenda;
};

export const normalizarAgendaFeriados = (payload: unknown): AgendaFeriadoConfig[] => {
  if (!Array.isArray(payload)) return [];

  const datas = new Set<string>();
  const itens: AgendaFeriadoConfig[] = [];

  for (const item of payload) {
    let dataRaw = "";
    let descricaoRaw = "";
    let tipoRaw = "local";

    if (typeof item === "string") {
      dataRaw = item.trim();
    } else if (typeof item === "object" && item !== null) {
      const row = item as Record<string, unknown>;
      dataRaw = String(row.data || "").trim();
      descricaoRaw = String(row.descricao || "").trim();
      tipoRaw = String(row.tipo || "local").trim().toLowerCase();
    }

    if (!/^\d{4}-\d{2}-\d{2}$/.test(dataRaw) || datas.has(dataRaw)) {
      continue;
    }

    datas.add(dataRaw);
    itens.push({
      data: dataRaw,
      descricao: descricaoRaw,
      tipo: tipoRaw === "nacional" ? "nacional" : "local",
    });
  }

  itens.sort((a, b) => a.data.localeCompare(b.data));
  return itens;
};

export const normalizarAgendaExcecoes = (payload: unknown): AgendaExcecaoConfig[] => {
  if (!Array.isArray(payload)) return [];

  const mapa = new Map<string, AgendaExcecaoConfig>();

  for (const item of payload) {
    if (typeof item !== "object" || item === null) continue;
    const row = item as Record<string, unknown>;

    const dataRaw = String(row.data || "").trim();
    if (!/^\d{4}-\d{2}-\d{2}$/.test(dataRaw)) {
      continue;
    }

    const ativo = typeof row.ativo === "boolean" ? row.ativo : true;
    const inicio = normalizarHoraHHMM(row.inicio, "08:00");
    const fim = normalizarHoraHHMM(row.fim, "18:00");
    const inicioMin = horarioParaMinutos(inicio) ?? 8 * 60;
    const fimMin = horarioParaMinutos(fim) ?? 18 * 60;
    const motivo = String(row.motivo || "").trim();

    mapa.set(dataRaw, {
      data: dataRaw,
      ativo,
      inicio: inicioMin < fimMin ? inicio : "08:00",
      fim: inicioMin < fimMin ? fim : "18:00",
      motivo,
    });
  }

  const itens = Array.from(mapa.values());
  itens.sort((a, b) => a.data.localeCompare(b.data));
  return itens;
};

export const obterExcecaoData = (
  dataIso: string,
  agendaExcecoes: AgendaExcecaoConfig[]
) => agendaExcecoes.find((item) => item.data === dataIso);

export const obterJornadaDia = (
  dataIso: string,
  agendaSemanal: AgendaSemanalConfig,
  agendaFeriados: AgendaFeriadoConfig[],
  agendaExcecoes: AgendaExcecaoConfig[] = []
): { fechado: boolean; inicio: string; fim: string; motivo: string } => {
  const excecao = obterExcecaoData(dataIso, agendaExcecoes);
  if (excecao) {
    if (!excecao.ativo) {
      return {
        fechado: true,
        inicio: excecao.inicio,
        fim: excecao.fim,
        motivo: excecao.motivo || "Agenda fechada por excecao",
      };
    }
    return {
      fechado: false,
      inicio: excecao.inicio,
      fim: excecao.fim,
      motivo: excecao.motivo ? `Excecao: ${excecao.motivo}` : "Horario especial desta data",
    };
  }

  const feriado = agendaFeriados.find((item) => item.data === dataIso);
  if (feriado) {
    const descricao = (feriado.descricao || "").trim();
    return {
      fechado: true,
      inicio: "00:00",
      fim: "00:00",
      motivo: descricao ? `Feriado: ${descricao}` : "Feriado",
    };
  }

  const data = new Date(`${dataIso}T00:00:00`);
  const diaSemanaJs = data.getDay(); // 0=Dom, 1=Seg...
  const diaKey: DiaSemanaKey = (diaSemanaJs === 0 ? "7" : String(diaSemanaJs)) as DiaSemanaKey;
  const cfg = agendaSemanal[diaKey] || DEFAULT_AGENDA_SEMANAL[diaKey];

  if (!cfg.ativo) {
    return {
      fechado: true,
      inicio: cfg.inicio,
      fim: cfg.fim,
      motivo: "Agenda fechada",
    };
  }

  return {
    fechado: false,
    inicio: cfg.inicio,
    fim: cfg.fim,
    motivo: "",
  };
};

export const slotDentroDaJornada = (
  hora: string,
  jornada: { fechado: boolean; inicio: string; fim: string }
) => {
  if (jornada.fechado) return false;
  const horaMin = horarioParaMinutos(hora);
  const inicioMin = horarioParaMinutos(jornada.inicio);
  const fimMin = horarioParaMinutos(jornada.fim);
  if (horaMin === null || inicioMin === null || fimMin === null) return false;
  return horaMin >= inicioMin && horaMin < fimMin;
};

export const validarHorarioAgendamento = (
  inicio: Date,
  fim: Date,
  agendaSemanal: AgendaSemanalConfig,
  agendaFeriados: AgendaFeriadoConfig[],
  agendaExcecoes: AgendaExcecaoConfig[] = []
): { valido: boolean; motivo: string } => {
  if (Number.isNaN(inicio.getTime()) || Number.isNaN(fim.getTime())) {
    return { valido: false, motivo: "Data ou hora invalida." };
  }
  if (fim.getTime() <= inicio.getTime()) {
    return { valido: false, motivo: "O horario final deve ser maior que o inicial." };
  }

  const dataInicio = toDateInput(inicio);
  const dataFim = toDateInput(fim);
  if (dataInicio !== dataFim) {
    return { valido: false, motivo: "O agendamento deve iniciar e terminar no mesmo dia." };
  }

  const jornada = obterJornadaDia(dataInicio, agendaSemanal, agendaFeriados, agendaExcecoes);
  if (jornada.fechado) {
    return { valido: false, motivo: jornada.motivo || "Agenda fechada neste dia." };
  }

  const inicioMin = inicio.getHours() * 60 + inicio.getMinutes();
  const fimMin = fim.getHours() * 60 + fim.getMinutes();
  const inicioJornadaMin = horarioParaMinutos(jornada.inicio);
  const fimJornadaMin = horarioParaMinutos(jornada.fim);

  if (inicioJornadaMin === null || fimJornadaMin === null) {
    return { valido: false, motivo: "Horario de funcionamento invalido nas configuracoes." };
  }

  if (inicioMin < inicioJornadaMin || fimMin > fimJornadaMin) {
    return {
      valido: false,
      motivo: `Horario fora do funcionamento da agenda (${jornada.inicio} as ${jornada.fim}).`,
    };
  }

  return { valido: true, motivo: "" };
};
